src/open-uri.c

```diff
@@ -689,29 +687,43 @@ handle_open_in_thread_func (GTask *task,
         {
           g_autofree char *real_path = get_real_path_for_doc_path (path, app_id);
 
-          if (open_uri->file_manager != NULL)
             {
               /* Try opening the directory via the file manager interface, then
                  fall back to a plain URI open */
               g_autoptr(GError) local_error = NULL;
               g_autoptr(GVariant) result = NULL;
               g_autoptr(GVariantBuilder) uris_builder = NULL;
               g_autofree char* item_uri = g_filename_to_uri (real_path, NULL, NULL);
+              g_autoptr(GDBusConnection) bus = NULL;
+
+              bus = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &local_error);
 
               uris_builder = g_variant_builder_new (G_VARIANT_TYPE ("as"));
               g_variant_builder_add (uris_builder, "s", item_uri);
 
-              result = g_dbus_proxy_call_sync (open_uri->file_manager,
-                                               FILE_MANAGER_SHOW_ITEMS,
-                                               g_variant_new ("(ass)", uris_builder, ""),
-                                               G_DBUS_CALL_FLAGS_NONE,
-                                               -1,
-                                               NULL,
-                                               &local_error);
+              if (bus)
+                result = g_dbus_connection_call_sync (bus,
+                                                      FILE_MANAGER_DBUS_NAME,
+                                                      FILE_MANAGER_DBUS_PATH,
+                                                      FILE_MANAGER_DBUS_IFACE,
+                                                      FILE_MANAGER_SHOW_ITEMS,
+                                                      g_variant_new ("(ass)", uris_builder, ""),
+                                                      NULL,   /* ignore returned type */
+                                                      G_DBUS_CALL_FLAGS_NONE,
+                                                      -1,
+                                                      NULL,
+                                                      &local_error);
+
               if (result == NULL)
                 {
-                  g_warning ("Failed to call " FILE_MANAGER_SHOW_ITEMS ": %s",
-                             local_error->message);
+                  if (g_error_matches (local_error, G_DBUS_ERROR,
```

Comment:
I wondered about that myself, but I think that would actually be wrong behaviour: if we can't contact the FileManager1 the first time, there's no reason it can't succeed when we try again later.

For example, if o.fd.FileManager1 isn't service-activatable, the first time we call it there might not be an implementation running, but by the time we ping it for a second time, the user might have started an implementation that grabbed the name.

At the moment, o.fd.FileManager1 implementations usually *are* service-activatable, but in some ways that's actually undesirable too, because D-Bus has no concept of a desktop-specific implementation of a bus name. So for example, if a GNOME user and an XFCE user share a computer, it's arbitrary (readdir() order, i.e. basically random) whether the session bus will choose GNOME's Nautilus or XFCE's Thunar to be today's implementation of o.fd.FileManager1.

At some point I'd like to resolve that by querying the `g_app_info_get_default_for_type()` for `inode/directory`, and then checking whether it's a `GDesktopAppInfo` that is `DBusActivatable`. If it is, we can try invoking *that* name via D-Bus as a FileManager1, before falling back to attempting service-activation - which will make us more likely to launch the user's preferred file manager. That's out of scope for solving the deadlock, though.

The other reason I didn't do this is that the OpenURI implementation runs in a worker thread, and there isn't any guarantee that we don't have two calls to OpenURI running concurrently; so we would need thread-synchronization between them.